### PR TITLE
refactor: replace `OnceLock` with `LazyLock` (round 2)

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?

Part of #11687

Replaced the remaining `OnceLock`s except for function [Documentation](https://github.com/apache/datafusion/blob/2464703c84c400a09cc59277018813f0e797bb4e/datafusion/functions/src/math/pi.rs#L89) instances.

## Rationale for this change
LazyLock provides a simpler way to define static variables.

## Are these changes tested?
Yes

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/core/tests/expr_api/mod.rs [hunk 1]
```diff
@@ -332,13 +330,13 @@ async fn evaluate_agg_test(expr: Expr, expected_lines: Vec<&str>) {
 /// Converts the `Expr` to a `PhysicalExpr`, evaluates it against the provided
 /// `RecordBatch` and compares the result to the expected result.
 fn evaluate_expr_test(expr: Expr, expected_lines: Vec<&str>) {
-    let batch = test_batch();
+    let batch = &TEST_BATCH;
```
Comment (jonahgao) [inline datafusion/core/tests/expr_api/mod.rs:333]: This change also avoids a clone.

File: datafusion/expr/src/test/function_stub.rs [hunk 2]
```diff
@@ -39,19 +39,14 @@ use crate::{
...
-            #[allow(non_upper_case_globals)]
-            static [< STATIC_ $UDAF >]: std::sync::OnceLock<std::sync::Arc<crate::AggregateUDF>> =
-                std::sync::OnceLock::new();
-
             #[doc = concat!("AggregateFunction that returns a [AggregateUDF](crate::AggregateUDF) for [`", stringify!($UDAF), "`]")]
             pub fn $AGGREGATE_UDF_FN() -> std::sync::Arc<crate::AggregateUDF> {
-                [< STATIC_ $UDAF >]
-                    .get_or_init(|| {
+                /// Singleton instance of [$UDAF], ensures the UDAF is only created once
+                static INSTANCE: std::sync::LazyLock<std::sync::Arc<crate::AggregateUDF>> =
```
Comment (jonahgao) [inline datafusion/expr/src/test/function_stub.rs:45]: Narrow the scope to inside the function, so we can use a simpler name.

File: datafusion/functions/src/macros.rs [hunk 3]
```diff
@@ -71,20 +71,18 @@ macro_rules! export_functions {
...
         pub fn $NAME() -> std::sync::Arc<datafusion_expr::ScalarUDF> {
-            $GNAME
-                .get_or_init(|| {
-                    std::sync::Arc::new(datafusion_expr::ScalarUDF::new_from_impl(
-                        <$UDF>::new(),
-                    ))
-                })
-                .clone()
+            /// Singleton instance of the function
+            static $GNAME: std::sync::LazyLock<
```
Comment (jonahgao) [inline datafusion/functions/src/macros.rs:78]: ` $GNAME` can be removed, maybe in a follow-up PR.

**Review by crepererum — APPROVED** [review]
thank you

**Pull request merged**